wasmlib/helpers/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+        if !self.err.is_null() {
+            let err = &*self.err;
+            let msg = ::std::slice::from_raw_parts(err.msg_ptr, err.msg_len.try_into().unwrap());
+            let msg_str = std::str::from_utf8(msg)?;
+            return Err(libipld::error::Error::msg(msg_str));
+        }
+        Ok(self.value)
+    }
+}
+
+#[repr(C)]
+pub struct BlockResp {
+    pub msg_len: u32,
+    pub msg_ptr: *const u8,
+    pub is_err: bool,
+}
+
+impl BlockResp {
+    /// Turn the data into a result which is either bytes or an error
+    ///
+    /// # Safety
+    ///
+    /// This function assumes each of the pointers are either nil or valid
+    pub unsafe fn to_result(&self) -> Result<&[u8], libipld::error::Error> {
+        let msg = ::std::slice::from_raw_parts(self.msg_ptr, self.msg_len.try_into().unwrap());
+        if self.is_err {
+            let msg_str = std::str::from_utf8(msg)?;
+            return Err(libipld::error::Error::msg(msg_str));
+        }
+        Ok(msg)
+    }
+}
+
+#[repr(C)]
+pub struct ByteWrapper {
+    pub msg_len: u32,
+    pub msg_ptr: *const u8,
+}
+
+#[repr(C)]
+pub struct ADLorWAC {
+    pub err: *const ByteWrapper,
+    pub adl_ptr: *const c_void,
+    pub adl_kind: u8,
+    pub wac: *const ByteWrapper,
+}
+
+#[repr(C)]
+pub struct IterResp {
+    pub err: *const ByteWrapper,
+    pub key: *const ByteWrapper,
+    pub value_adl_ptr: *const c_void,
+    pub value_adl_kind: u8,
+    pub val_wac: *const ByteWrapper,
+}
+
+#[repr(C)]
+pub struct ReadResp {
+    pub err: *const ByteWrapper,
+    pub bytes_read: u32,
+}
+
+#[repr(C)]
+pub struct BoolOrError {
+    pub err: *const ByteWrapper,
+    pub value: bool,
+}
+
+pub fn byte_vec_to_byte_wrapper(b: Vec<u8>) -> *const ByteWrapper {
+    let bx = b.into_boxed_slice();
+    let bx_len = bx.len() as u32;
+    let bytes_ptr = Box::into_raw(bx) as *const u8;
+
+    Box::into_raw(Box::new(ByteWrapper {
+        msg_len: bx_len,
+        msg_ptr: bytes_ptr,
+    }))
+}
+
+pub fn get_error(err: libipld::error::Error) -> *const ValueOrError {
+    let res = Box::new(ValueOrError {
+        value: std::ptr::null::<c_void>() as *mut c_void,
+        err: byte_vec_to_byte_wrapper(err.to_string().as_bytes().to_owned()),
+    });
+    Box::into_raw(res)
+}
+
+pub fn get_result_bytes(result: Result<Vec<u8>, libipld::error::Error>) -> *const ValueOrError {
+    match result {
+        Ok(v) => {
+            let bx = v.into_boxed_slice();
+            let bx_len = bx.len() as u32;
+            let bytes_ptr = Box::into_raw(bx) as *const u8;
+
+            let res = Box::new(ValueOrError {
+                err: std::ptr::null(),
+                value: Box::into_raw(Box::new(ByteWrapper {
+                    msg_len: bx_len,
+                    msg_ptr: bytes_ptr,
+                })) as *mut c_void,
+            });
+            Box::into_raw(res)
+        }
+        Err(err) => get_error(err),
+    }
+}
+
+// Extern Functions
+
+#[cfg(target_arch = "wasm32")]
+extern "C" {
```

Comment:
It'd be nice if the C ABI we expect hosts to implemented could be put in a separate document to make it instantly visible.

Reply:
Do you mean as text or as code? As code it's mostly in this file, and we can just move the few helper functions to a separate file from the type and function definitions.